feat: automate release process

mmagician opened the pull request:
For publishing Rust crates:
- dry-run publish on pushes to `next` & `main` (mostly a port of https://github.com/0xMiden/miden-base/blob/af318498f160a73e0f7ad51d2e9653bde2348281/.github/workflows/release-plz-dry-run.yml)
- publish to crates.io on a new GitHub release (mostly a port of https://github.com/0xMiden/miden-base/blob/07e493ae4022b314b995f26868b9ab5fd36abcce/.github/workflows/release-plz.yml)

For publishing web SDK package:
- publish to npm registry on a new GitHub release
- publish on each PR merged to next, if there was a `patch release` label on the PR
- in both cases, release proceeds only if the version differs from the previous version on main/next


closes https://github.com/0xMiden/miden-client/issues/1353

fkrause98 commented on .github/workflows/publish-web-client-next.yml:
```diff
@@ -1,68 +1,114 @@
+      - name: Verify version bump against next
+        id: check_version
+        run: |
+          set -eo pipefail
```
nit: consider also adding the 'u' switch to avoid undefined vars
```suggestion
          set -euo pipefail
```

dagarcia7 commented on .github/workflows/publish-web-client-next.yml:
```diff
@@ -1,68 +1,114 @@
+              echo "Unable to read crates/web-client/package.json from $BASE_SHA."
+              SHOULD_PUBLISH=false
+            fi
+          fi
```
I think this logic is sound already, but some thoughts I had that could potentially make it easier to read and slightly more performant:
* Change `SHOULD_PUBLISH` to default to `false` and flip after all all the checks pass. This seems safer in case there's any ambiguity or weirdness that happens in a step for whatever reason
* Add a short-circuit step that skips the `package.json` file read + comparison in case it didn't change at all compared to base via something like 

```
if ! git diff --name-only "$BASE_SHA"...HEAD -- crates/web-client/package.json | grep -q .; then
  echo "No changes to crates/web-client/package.json; skipping publish."
  exit 0
fi
```
so that the CI fails faster in case there are no changes
* If you decide to do `SHOULD_PUBLISH=false` by default, you can treat each step as a short circuit condition and call `exit 0` instead of changing the variable and having the code keep going through checking the logic gates

Anyway, I think the code works as is, just sharing an alternate approach if you think it makes sense! 👍

igamigo reviewed (approved):
Looks good. I still think the dry run might not work correcty (https://github.com/0xMiden/miden-client/pull/1403#discussion_r2470582243), but may be we can merge and debug later if needed?

mmagician commented on scripts/publish-crates.sh:
I just realized this seems to be [officially documented](https://doc.rust-lang.org/cargo/reference/config.html#hierarchical-structure) already:
> At present, when being invoked from a workspace, Cargo does not read config files from crates within the workspace

so it looks like a known limitation.
I think for now we need to have two calls to `publish` as I outlined above - which still makes it simpler than the current script, but not quite a one-liner.

mmagician commented:
> Looks good. I still think the dry run might not work correcty (https://github.com/0xMiden/miden-client/pull/1403#discussion_r2470582243), but may be we can merge and debug later if needed?

@igamigo thanks for testing this out manually! I now addressed it in [fix: split publishing in two steps](https://github.com/0xMiden/miden-client/pull/1403/commits/657bee6e42402888dd218f1d15990243c8c5c38b), and also tested locally to confirm that splitting it in two separate `publish` steps works.

Merging for now, and as you say we can iterate if needed.
